packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -360,111 +400,340 @@ contract AmmPool is IBlockReceiver {
+            if (ctx.ammBalances[i] > ctx.ammBalancesInAccount[i]) {
+                uint amount = ctx.ammBalances[i] - ctx.ammBalancesInAccount[i];
+                processDeposit(ctx, ctx.tokens[i], amount);
+            } else if (ctx.ammBalancesInAccount[i] > ctx.ammBalances[i]) {
+                uint amount = ctx.ammBalancesInAccount[i] - ctx.ammBalances[i];
+                processWithdrawal(ctx, ctx.tokens[i], amount);
+            }
+        }
+
+        // The ending AMM updates
+        processAmmUpdates(ctx, false);
+
+        emit QueueItemsProcessed(poolTransactions.length);
+        return ctx.numTransactionsConsumed;
+    }
+
+    function totalSupply()
+        public
+        view
+        returns (uint256)
+    {
+        return poolSupply;
+    }
+
+    function mint(address owner, uint amount)
+        internal
+    {
+        poolSupply = poolSupply.add(amount);
+        poolBalance[owner] = poolBalance[owner].add(amount);
+    }
+
+    function burn(address owner, uint amount)
+        internal
+    {
+        poolSupply = poolSupply.sub(amount);
+        poolBalance[owner] = poolBalance[owner].sub(amount);
+    }
+
+    function processAmmUpdates(
+        Context memory ctx,
+        bool   start
+        )
+        internal
+    {
+        for (uint i = 0; i < ctx.tokens.length; i++) {
             // Check that the AMM update in the block matches the expected update
-            AmmUpdateTransaction.AmmUpdate memory update = _block.readAmmUpdate(txIdx++);
+            AmmUpdateTransaction.AmmUpdate memory update = ctx._block.readAmmUpdate(ctx.txIdx++);
             require(update.owner == address(this), "INVALID_TX_DATA");
             require(update.accountID == accountID, "INVALID_TX_DATA");
-            require(update.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
+            require(update.tokenID == ctx.tokens[i].tokenID, "INVALID_TX_DATA");
             require(update.feeBips == feeBips, "INVALID_TX_DATA");
-            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            require(update.tokenWeight == (start ? 0 : ctx.tokens[i].weight), "INVALID_TX_DATA");
             // Now approve this AMM update
             update.validUntil = 0xffffffff;
-            bytes32 txHash = AmmUpdateTransaction.hashTx(exchangeDomainSeparator, update);
+            bytes32 txHash = AmmUpdateTransaction.hashTx(ctx.exchangeDomainSeparator, update);
             exchange.approveTransaction(address(this), txHash);
-            numTransactionsConsumed++;
-            // AMM account balance now available onchain
-            ammBalancesInAccount[i] = update.balance;
-            ammBalances[i] = ammBalancesInAccount[i];
+            ctx.numTransactionsConsumed++;
+            if (start) {
+                // AMM account balance now available onchain
+                ctx.ammBalancesInAccount[i] = update.balance;
+                ctx.ammBalances[i] = ctx.ammBalancesInAccount[i];
```

Comment:
> the new weights are correct

The weights always come directly from the contract itself so can's see additional checks that can be added for this.

> ctx.ammBalancesInAccount[i] == ctx.ammBalances[i]?

Yep, already doing that now! (additional safety check because transfers can have small rounding errors).

Reply:
> The weights always come directly from the contract itself
Sorry, I thought the new weights are calculated by the operator and the contract just verify the new weights based on the new balances. 

> Yep, already doing that now!
Do you mean the code will be pushed to this PR, but not here yet?

Reply:
> Do you mean the code will be pushed to this PR, but not here yet?

Yes, will push shortly.